precompiles/staking/method_distribute_test.go

```diff
@@ -0,0 +1,220 @@
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(1000))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(0)}...,
+		)
+
+		// Call method.
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid token amount: 0")
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
+		require.False(t, ok)
+	})
+
+	t.Run("should fail to distribute more than allowed to staking", func(t *testing.T) {
+		// Setup test.
+		s := newTestSuite(t)
+
+		// Set caller balance.
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(999))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(1000)}...,
+		)
+
+		// Call method.
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "invalid allowance, got 999, wanted 1000")
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
+		require.False(t, ok)
+	})
+
+	t.Run("should fail to distribute more than user balance", func(t *testing.T) {
+		// Setup test.
+		s := newTestSuite(t)
+
+		// Set caller balance.
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(100000))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(1001)}...,
+		)
+
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "execution reverted")
+
+		// Unpack and check result boolean.
+		res, err := s.methodID.Outputs.Unpack(success)
+		require.NoError(t, err)
+
+		ok := res[0].(bool)
+		require.False(t, ok)
+	})
+
+	t.Run("should distribute and lock ZRC20", func(t *testing.T) {
+		// Setup test.
+		s := newTestSuite(t)
+
+		// Set caller balance.
+		s.fungibleKeeper.DepositZRC20(s.ctx, s.zrc20Address, s.defaultCaller, big.NewInt(1000))
+
+		// Allow staking to spend ZRC20 tokens.
+		allowStaking(t, s, big.NewInt(1000))
+
+		// Setup method input.
+		s.mockVMContract.Input = packInputArgs(
+			t,
+			s.methodID,
+			[]interface{}{s.zrc20Address, big.NewInt(1000)}...,
+		)
+
+		success, err := s.contract.Run(s.mockEVM, s.mockVMContract, false)
+
+		// Check error.
+		require.NoError(t, err)
```

Comment:
Can we check fee collector balance ?

Reply:
Added in 4eae5e2fd58f8c7f16e324ea8321dd12944c563b.
Anyway I'm under the impression that mocking the state is not completely working, that's why I didn't add this check.
In the fungible lock/unlock API the transfer and lock of ZRC20 assets works well, so do the same in e2e which use the API, such as distribute and bank.
However, for some reason in the unit test for method distribute it doesn't work as expected and the state doesn't reflect properly the changes.

Reply:
Okay , 
so I guess something like this 
```
	zrc20Denom := precompiletypes.ZRC20ToCosmosDenom(s.zrc20Address)
		balance, err := s.sdkKeepers.BankKeeper.Balance(s.ctx, &banktypes.QueryBalanceRequest{
			Address: feeCollectorAddress,
			Denom:   zrc20Denom,
		})
		require.NoError(t, err)
		require.NotEqual(t, uint64(0), balance.Balance.Amount.Uint64())
```

wont work as the balance is always 0 no matter if its successful or failed ?

Reply:
That's correct.